app/cmd/start.go

```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
 
 	// Start proxy server
-	proxyGRPCServer, proxyGRPCListener, err := setupProxyGRPCServer(logsDir, proxyServiceAddress, diskServiceAddress, spdkServiceAddress, tlsConfig, shutdownCh)
+	proxyGRPCServer, proxyGRPCListener, err := setupProxyGRPCServer(logsDir, proxyServiceAddress, diskServiceAddress, spdkServiceAddress, tlsConfig, shutdown.ch)
 	if err != nil {
 		return err
 	}
 	go func() {
+		defer func() {
+			shutdown.once.Do(func() {
+				close(shutdown.ch)
+			})
+			wg.Done()
+			logrus.Info("Stopped proxy gRPC server")
+		}()
+
 		if err := proxyGRPCServer.Serve(proxyGRPCListener); err != nil {
 			logrus.WithError(err).Error("Stopping proxy gRPC server")
 		}
-		// graceful shutdown before exit
-		close(shutdownCh)
 	}()
 	logrus.Infof("Instance Manager proxy gRPC server listening to %v", proxyServiceAddress)
 
 	// Start process manager server
-	pm, pmGRPCServer, pmGRPCListener, err := setupProcessManagerGRPCServer(processPortRange, logsDir, processManagerServiceAddress, shutdownCh)
+	pm, pmGRPCServer, pmGRPCListener, err := setupProcessManagerGRPCServer(processPortRange, logsDir, processManagerServiceAddress, shutdown.ch)
 	if err != nil {
 		return err
 	}
 	go func() {
+		defer func() {
+			shutdown.once.Do(func() {
+				close(shutdown.ch)
+			})
+			wg.Done()
+			logrus.Info("Stopped process manager gRPC server")
+		}()
+
 		if err := pmGRPCServer.Serve(pmGRPCListener); err != nil {
 			logrus.WithError(err).Error("Stopping process manager gRPC server")
 		}
-		// graceful shutdown before exit
 		cleanup(pm)
-		close(shutdownCh)
 	}()
 	logrus.Infof("Instance Manager process manager gRPC server listening to %v", listen)
 
 	// Start SPDK server
+	var spdkGRPCServer *grpc.Server
+	var spdkGRPCListener net.Listener
 	if spdkEnabled {
-		spdkGRPCServer, spdkGRPCListener, err := setupSPDKGRPCServer(spdkPortRange, spdkServiceAddress, shutdownCh)
+		spdkGRPCServer, spdkGRPCListener, err = setupSPDKGRPCServer(spdkPortRange, spdkServiceAddress, shutdown.ch)
 		if err != nil {
 			return err
 		}
 		go func() {
+			defer func() {
+				shutdown.once.Do(func() {
+					close(shutdown.ch)
+				})
+				wg.Done()
+				logrus.Info("Stopped SPDK gRPC server")
+			}()
+
 			if err := spdkGRPCServer.Serve(spdkGRPCListener); err != nil {
 				logrus.WithError(err).Error("Stopping SPDK gRPC server")
 			}
-			// graceful shutdown before exit
-			close(shutdownCh)
 		}()
 		logrus.Infof("Instance Manager SPDK gRPC server listening to %v", spdkServiceAddress)
 	}
 
+	// Register signal handler
 	sigs := make(chan os.Signal, 1)
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
 	go func() {
 		sig := <-sigs
+
 		logrus.Infof("Instance Manager received %v to exit", sig)
+		instanceGRPCServer.Stop()
+		if spdkGRPCServer != nil {
+			spdkGRPCServer.Stop()
+		}
 		pmGRPCServer.Stop()
+		proxyGRPCServer.Stop()
+		diskGRPCServer.Stop()
 	}()
 
-	return <-shutdownCh
+	wg.Wait()
+
+	if spdkEnabled {
+		stopSPDKTgt(30 * time.Second)
```

Comment:
We start `spdk_tgt` in longhorn-spdk-engine, so can we do stop there as well to make responsibility segregation? The lifecycle of it should be together with the `spdkGRPCServer`. Is this feasible?

Reply:
We start spdk_tgt before starting longhorn-instance-manager ([code](https://github.com/longhorn/longhorn-instance-manager/blob/master/package/instance-manager#L63)). longhorn-spdk-engine ensures the health of spdk_tgt and restarts it if it is somehow crashed.

Therefore, stopping spdk_tgt before exiting longhorn-instance-manager looks reasonable. WDYT?

Reply:
Sounds good.